File: scala/leet_code/42.scala
/** This module provides a solution to the problem of trapping rain water.
  * Given an array of integers representing an elevation map where the width of each bar is 1,
  * compute how much water it can trap after raining.
  */
object Solution {

  /** Function to calculate the total amount of water that can be trapped between the bars.
    *
    * @param height an Array[Int] representing the heights of the bars.
    * @return an Int representing the total amount of water that can be trapped.
    */
  def trap(height: Array[Int]): Int = {
    // If the height array is empty, return 0
    if (height.isEmpty) 0
    else {
      var left = 0
      var right = height.length - 1
      var leftMax = 0
      var rightMax = 0
      var trappedWater = 0

      // Iterate until the left pointer is less than the right pointer
      while (left < right) {
        // If the height of the bar on the left is less than the height of the bar on the right
        if (height(left) < height(right)) {
          // If the height of the bar on the left is greater than or equal to the maximum height on the left
          if (height(left) >= leftMax) {
            // Update the maximum height on the left
            leftMax = height(left)
          } else {
            // Add the difference between the maximum height on the left and the height of the bar on the left to the trappedWater
            trappedWater += leftMax - height(left)
          }
          // Move the left pointer to the right
          left += 1
        } else {
          // If the height of the bar on the right is greater than or equal to the maximum height on the right
          if (height(right) >= rightMax) {
            // Update the maximum height on the right
            rightMax = height(right)
          } else {
            // Add the difference between the maximum height on the right and the height of the bar on the right to the trappedWater
            trappedWater += rightMax - height(right)
          }
          // Move the right pointer to the left
          right -= 1
        }
      }

      // Return the trappedWater
      trappedWater
    }
  }
}

File: swift/leet_code/42.swift
/**
 A `Solution` class that provides a method to calculate the trapping rain water.

 This class provides a method `trap` which calculates how much rain water can be trapped between the bars, given the heights of the bars. The heights of the bars are represented as an array of integers.

 - Note: The `trap` method uses a two-pointer approach to solve the problem.

 - Complexity: This method runs in O(n) time, where n is the number of bars. It uses O(1) extra space.
 */
class Solution {
    /**
     Calculate the amount of trapped rain water.

     This method calculates and returns the total amount of rain water that can be trapped, given the heights of the bars.

     - Parameter height: An array of integers where each integer represents the height of a bar.
     - Returns: The total amount of trapped rain water.

     - Note: If the height array is empty, this method will return 0.
     */
    func trap(_ height: [Int]) -> Int {
        var leftMax = 0 // Maximum height of the bar from the left up to the current position
        var rightMax = 0 // Maximum height of the bar from the right up to the current position
        var lp = 0 // Left pointer
        var rp = height.count - 1 // Right pointer
        var result = 0 // Resultant trapped water

        // Loop until the two pointers meet
        while lp < rp {
            if height[lp] <= height[rp] {
                if height[lp] >= leftMax {
                    leftMax = height[lp] // Update the left maximum
                } else {
                    result += leftMax - height[lp] // Calculate the trapped water
                }
                lp += 1 // Move the left pointer
            } else {
                if height[rp] >= rightMax {
                    rightMax = height[rp] // Update the right maximum
                } else {
                    result += rightMax - height[rp] // Calculate the trapped water
                }
                rp -= 1 // Move the right pointer
            }
        }
        return result // Return the total trapped water
    }
}
